pkg/services/authn/clients/jwt.go

```diff
@@ -0,0 +1,208 @@
+package clients
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/auth"
+	"github.com/grafana/grafana/pkg/services/authn"
+	"github.com/grafana/grafana/pkg/services/login"
+	"github.com/grafana/grafana/pkg/services/org"
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/grafana/grafana/pkg/util/errutil"
+	"github.com/jmespath/go-jmespath"
+)
+
+var _ authn.Client = new(JWT)
+
+var (
+	ErrJWTInvalid = errutil.NewBase(errutil.StatusUnauthorized,
+		"jwt.invalid", errutil.WithPublicMessage("Failed to verify JWT"))
+	ErrJWTMissingClaim = errutil.NewBase(errutil.StatusUnauthorized,
+		"jwt.missing_claim", errutil.WithPublicMessage("Missing mandatory claim in JWT"))
+	ErrJWTInvalidRole = errutil.NewBase(errutil.StatusForbidden,
+		"jwt.invalid_role", errutil.WithPublicMessage("Invalid Role in claim"))
+)
+
+func ProvideJWT(jwtService auth.JWTVerifierService, cfg *setting.Cfg) *JWT {
+	return &JWT{
+		cfg:        cfg,
+		log:        log.New(authn.ClientJWT),
+		jwtService: jwtService,
+	}
+}
+
+type JWT struct {
+	cfg        *setting.Cfg
+	log        log.Logger
+	jwtService auth.JWTVerifierService
+}
+
+func (s *JWT) Authenticate(ctx context.Context, r *authn.Request) (*authn.Identity, error) {
+	jwtToken := r.HTTPRequest.Header.Get(s.cfg.JWTAuthHeaderName)
+	if jwtToken == "" && s.cfg.JWTAuthURLLogin {
+		jwtToken = r.HTTPRequest.URL.Query().Get("auth_token")
+	}
+
+	// Strip the 'Bearer' prefix if it exists.
+	jwtToken = strings.TrimPrefix(jwtToken, "Bearer ")
+
+	claims, err := s.jwtService.Verify(ctx, jwtToken)
+	if err != nil {
+		s.log.Debug("Failed to verify JWT", "error", err)
+		return nil, ErrJWTInvalid.Errorf("failed to verify JWT")
```

Comment:
Should we wrap error from jwt service?

Reply:
The previous implementation didn't do it, I'm not afraid a lot of information is being given to an attacker, we could change it